### PR TITLE
Add support for array of enums

### PR DESCRIPTION
**Pull request by sanderdlm**
Allow the use of 'array' and 'simple_array' types in combination with the enumType attribute parameter.

This is my first PR to the project, any feedback or instructions are welcome.

I had to duplicate the try/catch which introduces a bit of redundancy. Should I make a new method for this?

### CODE REVIEW

**Comment by derrabus** [comment]
Thank you for your PR. As this is a new feature, please target the 2.12.x branch. Also, we need tests that cover your change.

**Comment by beberlei** [comment]
This is a great idea, i will comment on some code improvements later

**Comment by derrabus** [comment]
By the way, you can run all those complaining tools locally and get faster feedback. You don't have to push your changes through our CI everytime.

Once you've installed the dependencies with `composer update`, the following tools should be available:

* `./vendor/bin/phpcs`: PHP CodeSniffer, the tool that performs our code-style check.
* `./vendor/bin/phpcbf`: PHP Code Beautifier, a tool that can fix most PHPCS errors automatically
* `./vendor/bin/phpstan`: PHPStan (static code analysis)
* `./vendor/bin/psalm`: Psalm (the other static code analyzer)

**Comment by sanderdlm** [comment]
Cheers, think I've cleared them all now :slightly_smiling_face:

**Comment by dreadnip** [comment]
I've noticed an issue while testing.

Using the `array` type, this solutions works, and enums are correctly serialized into the database:
```a:2:{i:0;E:23:"App\Enum\Weekday:Monday";i:1;E:23:"App\Enum\Weekday:Friday";}```

Using the `simple_array` type, the enum does not pass the `implode(',', $value)` step:
`Object of class App\Enum\Weekday could not be converted to string`

Maybe another type called `enum_array` could be introduced that would check for the content of the array and call `->value` on the members before imploding? Or this could be built into `simple_array`?

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php [hunk 1]
```diff
@@ -67,18 +75,37 @@ public function setValue($object, $value = null): void
...
-
-                throw MappingException::invalidEnumValue(
-                    get_class($object),
-                    $this->originalReflectionProperty->getName(),
-                    (string) $value,
-                    $enumType,
-                    $e
-                );
+
+            if (is_array($value)) {
```
Comment (beberlei) [inline lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php:74]: What do you think of refactoring it to have less duplication?

Yo could introduce a private helper method `initializeEnumValue` that extracts what is currently in line 81-94 and 96-108. I believe its exactly the same code or did I miss something?

**Review by beberlei — CHANGES_REQUESTED** [review]
Initially I was wondering if the generic `is_array` or else handling isn't a big "dangerous", because there is no connection to the `Type` instance, where we know if value is an array or not, essentially allowing to "use this wrong".

But since getValue/setValue only happen to be used on data coming from the database, not from user input, I haven't found a simple example where this is not programmers error.

I am inclined to ignore that for now and see what is coming out of this change.

Very happy about the change, its simple addition allowing arrays of enums makes sense.

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 1 above)
Comment (sanderdlm) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php:74]: The code is exactly the same yeah. I've added the helper in [01a1b75](https://github.com/doctrine/orm/pull/9497/commits/01a1b75086982bed44c1be46c40aeca1f613b774).

The only thing I'm not sure about is the wording of '$item', but I couldn't come up with anything better.

**Comment by derrabus** [comment]
How can we move forward here? The PR needs a rebase at least.

**Comment by sanderdlm** [comment]
I can't replicate the phpcs failure locally.

**Comment by derrabus** [comment]
> I can't replicate the phpcs failure locally.

I've pushed a commit that should fix it.

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php [hunk 2]
```diff
@@ -56,6 +58,12 @@ public function getValue($object = null)
             return null;
         }
 
+        if (is_array($enum)) {
+            return array_map(static function ($item): mixed {
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]:
```suggestion
            return array_map(static function (BackedEnum $item): mixed {
```

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php [hunk 3]
```diff
@@ -66,22 +74,38 @@ public function getValue($object = null)
...
+                $value = $this->initializeEnumValue($object, $value);
             }
         }
 
         $this->originalReflectionProperty->setValue($object, $value);
     }
+
+    /**
+     * @param object $object
+     * @param mixed  $value
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]:
```suggestion
     * @param int|string $value
```

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php [hunk 4]
```diff
@@ -66,22 +74,38 @@ public function getValue($object = null)
...
-
-                throw MappingException::invalidEnumValue(
-                    get_class($object),
-                    $this->originalReflectionProperty->getName(),
-                    (string) $value,
-                    $enumType,
-                    $e
-                );
+            if (is_array($value)) {
+                $value = array_map(function ($item) use ($object): mixed {
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]:
```suggestion
                $value = array_map(function ($item) use ($object): BackedEnum {
```

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 2 above)
Comment (sanderdlm) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Changed in [b6e7472](https://github.com/doctrine/orm/pull/9497/commits/b6e7472f91ffe29851eb08093f6f8c1e8eb817a8)

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 3 above)
Comment (sanderdlm) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Changed in [b6e7472](https://github.com/doctrine/orm/pull/9497/commits/b6e7472f91ffe29851eb08093f6f8c1e8eb817a8)

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 4 above)
Comment (sanderdlm) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Changed in [b6e7472](https://github.com/doctrine/orm/pull/9497/commits/b6e7472f91ffe29851eb08093f6f8c1e8eb817a8)

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 3 above)
Comment (sanderdlm) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Oh, the parameter should be `int|string|array`, no?

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php [hunk 5]
```diff
@@ -56,32 +58,54 @@ public function getValue($object = null)
...
+        }
+
         return $enum->value;
     }
 
     /**
-     * @param object $object
-     * @param mixed  $value
+     * @param object     $object
+     * @param int|string $value
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: This should silence PHPStan:

```suggestion
     * @param object                    $object
     * @param int|string|int[]|string[] $value
```

File: lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php (hunk 3 above)
Comment (derrabus) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Yes, my bad.

Comment (derrabus) [reply lib/Doctrine/ORM/Mapping/ReflectionEnumProperty.php]: Actually no, `int|string` is correct here.

**Review by derrabus — DISMISSED** [review]

**Comment by derrabus** [comment]
Thank you @dreadnip!

**Review by derrabus — APPROVED** [review]

**Pull request merged**